source/common/common/utility.cc

```diff
@@ -20,8 +21,77 @@
 #include "spdlog/spdlog.h"
 
 namespace Envoy {
+
 std::string DateFormatter::fromTime(const SystemTime& time) const {
-  return fromTime(std::chrono::system_clock::to_time_t(time));
+  const std::string new_format_string = setCustomField(
+      // "%f" is the custom field for subsecond.
+      CustomFields::f(),
+      fmt::FormatInt(
+          // By default we want nanoseconds.
+          std::chrono::duration_cast<std::chrono::nanoseconds>(time.time_since_epoch()).count())
+          .str()
+          .substr(10));
+
+  const time_t current_time = std::chrono::system_clock::to_time_t(time);
+  if (new_format_string.empty()) {
+    return fromTime(current_time);
+  }
+
+  // Send the newly formatted format string to stftime.
+  return fromTime(current_time, new_format_string);
+}
+
+std::string DateFormatter::fromTime(time_t time, const std::string& new_format_string) const {
+  tm current_tm;
+  gmtime_r(&time, &current_tm);
+
+  std::array<char, 1024> buf;
+  strftime(&buf[0], buf.size(), new_format_string.c_str(), &current_tm);
+  return std::string(&buf[0]);
+}
+
+std::string DateFormatter::setCustomField(const std::string& field,
+                                          const std::string& value) const {
+  size_t found = format_string_.find(field);
+  if (found == std::string::npos) {
+    // If no field specifier inside the current format string, return an empty string and
+    // subsequently use the default formatter.
+    return EMPTY_STRING;
+  }
+
+  size_t index = found;
+  std::string new_format_string = format_string_;
+
+  while (found != std::string::npos) {
+    index = found;
+
+    size_t start = found - 1;
+    if (start < new_format_string.size()) {
+      std::string sub = new_format_string.substr(start, 2);
+      if (sub.at(0) == '%' && sub.size() == 2) {
+        // This field definitely has no width specifier, e.g. %f.
+        new_format_string = new_format_string.replace(start, 2, value);
+      } else {
+        start = found - 2;
+
+        // This field potentially has a valid width specifier, e.g. %3f.
+        if (start < new_format_string.size()) {
+          std::string sub = new_format_string.substr(start, 2);
+
+          if (sub.at(0) == '%' && sub.size() == 2) {
+            uint64_t width;
+            if (StringUtil::atoul(sub.substr(1, 1).c_str(), width)) {
+              new_format_string = new_format_string.replace(start, 3, value.substr(0, width));
```

Comment:
This is pretty complex, could we use `std::regex` (`replace`?) to handle this instead?

Reply:
Sure.

Reply:
@htuch I tried to compare the proposed approach with the regex one (both has no caching impl yet) for rendering `%s%f%3f`:

```
-------------------------------------------------------------------------------------
Benchmark                                              Time           CPU Iterations
-------------------------------------------------------------------------------------
BM_DateTimeFormatter                                3783 ns       3742 ns     204552
BM_DateTimeFormatterRegex                          16500 ns      16449 ns      41048
```

The code for rendering value that uses regex:

```cpp
std::string DateFormatter::setCustomFieldRegex(const std::regex& field,
                                               const std::string& value) const {
  std::string new_format_string = format_string_;
  std::smatch result;
  while (regex_search(new_format_string, result, field)) {
    ASSERT(result.size() == 2);
    const std::string& width_specifier = result[1];
    const uint32_t width = width_specifier.empty() ? value.size() : std::stoul(result[1]);
    new_format_string.replace(result.position(), result.length(), value.substr(0, width));
  }
  return new_format_string;
}
```

Reply:
I agree, regex is probable infeasible given > 4x slowdown. I think there might be a way to use regexes in a performant way; i.e. first use them to decompose the format string into a structured object and then populate that dynamically during the request. It's probably not worth doing in this initial implementation.

Reply:
@ggreenway points out that this could be more easily optimized by just precompiling the regex. Can you benchmark again with this?

Reply:
Do you mean by (statically) initializing it before use? e.g. https://github.com/dio/envoy/blob/3dec8584bd905a556daf28b6f725eccdee35a39d/source/common/common/utility.cc#L31? That's what I did, but will explore more.

Reply:
Yeah, constructing the `std::regex` should compile it. `SUBSECOND_PATTERN` is super simple, it seems that if regex is 4x slower then the built-in regexes library performance is pretty bad, I wonder if we could consider a third-party lib...

Reply:
Can you try with flag std::regex::optimize? See http://en.cppreference.com/w/cpp/regex/basic_regex

Reply:
@htuch @ggreenway I tried, but if we blindly search and replace at runtime it will be somewhat slower.

I decided to do parsing whenever constructing a `DateFormatter` object. This parsing step gives a list of position and width of each subsecond.